File: clojure/src/advent2016/day1.clj
(ns advent2016.day1
  (:require [clojure.string :as s]))


(def example1 "R2, L3")
(def example2 "R2, R2, R2")
(def example3 "R5, L5, R5, R3")
(def puzzle (vec (line-seq (clojure.java.io/reader "../day1.data"))))


(defn walk-one
  "Travel one position in the given heading."
  [heading [x y]]
  (condp = heading
    :n [x (dec y)]
    :s [x (inc y)]
    :w [(dec x) y]
    :e [(inc x) y]))


(defn walk
  "Given a heading, an instruction and a starting position, determine all the
  intermediate positions, including the end position."
  [heading [dir count] pos]
  (let [heading (rotate heading dir)]
    [heading
     (->> (iterate #(walk-one heading %) pos)
          (take (inc count))
          (drop 1)
          vec)]))


(defn rotate
  "Given a heading and an instruction direction, determine the new heading."
  [heading dir]
  (case dir
    :L (case heading
         :n :w
         :w :s
         :s :e
         :e :n)
    :R (case heading
         :n :e
         :e :s
         :s :w
         :w :n)))


(defn travel
  "Follow the instructions from a starting position.

  The result is a vector of each intermediate position travelled, including the
  end position."
  [pos instructions]
  (->> instructions
       (reductions (fn [[heading steps] instruction]
                     (walk heading instruction (peek steps)))
                   [:n [pos]])
       (mapcat second)
       vec))


(defn distance
  "Calculate the taxicab distance between two points."
  [[sx sy] [ex ey]]
  (+ (Math/abs (- ex sx))
     (Math/abs (- ey sy))))


(defn translate-instruction
  "Translate a string instruction such as \"R2\" into `[:R 2]`."
  [instruction]
  [(keyword (subs instruction 0 1))
   (Integer/parseInt (subs instruction 1))])


(defn first-duplicate
  "Find the first duplicate value in a collection."
  [coll]
  (loop [seen     #{}
         [x & xs] coll]
    (cond
      (seen x)    x
      (empty? xs) nil
      :else       (recur (conj seen x) xs))))


(defn solve
  "Solve a puzzle given input instructions as a string."
  [s]
  (let [pos          [0 0]
        instructions (map translate-instruction (s/split s #", "))
        trail        (travel pos instructions)]
    (distance pos
              (or (first-duplicate trail)
                  (peek trail)))))

File: python/day1.py
def new_heading(prev_heading, dir):
    return {
        (None, 'R'): 'E',
        ('N', 'R'): 'E',
        ('S', 'R'): 'W',
        ('E', 'R'): 'S',
        ('W', 'R'): 'N',
        (None, 'L'): 'W',
        ('N', 'L'): 'W',
        ('S', 'L'): 'E',
        ('E', 'L'): 'N',
        ('W', 'L'): 'S'}[(prev_heading, dir)]


def walk(heading, count, (x, y)):
    if heading == 'N': return (0, -count)
    elif heading == 'S': return (0, +count)
    elif heading == 'W': return (-count, 0)
    elif heading == 'E': return (+count, 0)
    raise NotImplementedError()


def transform_directions(s):
    return [(x[0], int(x[1:])) for x in s.split(', ')]


def update_seen(seen, (ox, oy), (dx, dy)):
    # Fill in the intermediate positions between a start and an end.
    for i in xrange(0, dx, -1 if dx < 0 else 1):
        pos = (ox + i, oy)
        if pos in seen:
            return pos
        seen.append(pos)
    for i in xrange(0, dy, -1 if dy < 0 else 1):
        pos = (ox, oy + i)
        if pos in seen:
            return pos
        seen.append(pos)
    return None


def travel(directions):
    pos = (0, 0)
    heading = None
    seen = []
    for step, (dir, count) in enumerate(transform_directions(directions), 1):
        #old_pos = pos
        heading = new_heading(heading, dir)
        paces = walk(heading, count, pos)
        seen_pos = update_seen(seen, pos, paces)
        if seen_pos is not None:
            #print 'twice!', seen_pos
            return seen_pos
        pos = (pos[0] + paces[0], pos[1] + paces[1])
        #print step, '{}{}'.format(dir, count), old_pos, '->', pos, heading
    return pos


def calculate_blocks(end, start=(0, 0)):
    return abs(end[0] - start[0]) + abs(end[1] - start[1])


example1 = 'R2, L3'
example2 = 'R2, R2, R2'
example3 = 'R5, L5, R5, R3'
example4 = 'R4, R3, R5, L3, L5, R2, L2, R5, L2, R5, R5, R5, R1, R3, L2, L2, L1'
puzzle = 'R4, R3, R5, L3, L5, R2, L2, R5, L2, R5, R5, R5, R1, R3, L2, L2, L1, R5, L3, R1, L2, R1, L3, L5, L1, R3, L4, R2, R4, L3, L1, R4, L4, R3, L5, L3, R188, R4, L1, R48, L5, R4, R71, R3, L2, R188, L3, R2, L3, R3, L5, L1, R1, L2, L4, L2, R5, L3, R3, R3, R4, L3, L4, R5, L4, L4, R3, R4, L4, R1, L3, L1, L1, R4, R1, L4, R1, L1, L3, R2, L2, R2, L1, R5, R3, R4, L5, R2, R5, L5, R1, R2, L1, L3, R3, R1, R3, L4, R4, L4, L1, R1, L2, L2, L4, R1, L3, R4, L2, R3, L1, L5, R4, R5, R2, R5, R1, R5, R1, R3, L3, L2, L2, L5, R2, L2, R5, R5, L2, R3, L5, R5, L2, R4, R2, L1, R3, L5, R3, R2, R5, L1, R3, L2, R2, R1'


if __name__ == '__main__':
    print calculate_blocks(travel(puzzle))
